File: Python/Problems/Problems/problem1911.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem1911(Problem):
    def test(self) -> bool:
        nums = [6, 2, 1, 2, 4, 5]

        expected = 10

        result = self.maxAlternatingSum(nums)

        print(result)

        return result == expected

    def maxAlternatingSum(self, nums: List[int]) -> int:
        n = len(nums)
        even = 0
        odd = nums[0]
        for i in range(1, n):
            even = max(even, odd - nums[i])
            odd = max(odd, even + nums[i])


        return max(even, odd)

File: C++/Problems/Problems/problem_1911.cpp
#include "problem_1911.h"

bool problem_1911::test() {
    std::vector nums { 6, 2, 1, 2, 4, 5 };

    int expected = 10;

    auto result = maxAlternatingSum(nums);

    std::cout << result << std::endl;

    return result == expected;
}

long long problem_1911::maxAlternatingSum(std::vector<int> &nums) {
    int n = nums.size();
    long long even = 0;
    long long odd = nums[0];
    for (int i = 1; i < n; ++i)
    {
        even = std::max(even, odd - nums[i]);
        odd = std::max(odd, even + nums[i]);
    }


    return std::max(even, odd);
}
